### PR TITLE
MGMT-9227: Create deploy_capi_cluster script for CAPI CI

### PR DESCRIPTION
**Pull request by mkowalski**
This PR creates a script responsible for deploying a cluster using CAPI
in the setup with Infrastructure Operator. It creates a required
InfraEnv, leverages BareMetalHost created by dev-scripts and later
installs HyperShift to install a cluster.

It is meant to be used by the e2e-metal-assisted-operator-capi job.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [x] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @javipolo 
/cc @eranco74 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: deploy/operator/capi/deploy_capi_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env bash
+
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+__root="$(realpath ${__dir}/../../..)"
+source ${__dir}/../common.sh
+source ${__dir}/../utils.sh
+source ${__dir}/none_platform_utils.sh
```
Comment (vrutkovs) [inline deploy/operator/capi/deploy_capi_cluster.sh:7]:
```
./deploy_capi_cluster.sh: line 7: /home/assisted-service/deploy/operator/capi/none_platform_utils.sh: No such file or directory
```
in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3690/pull-ci-openshift-assisted-service-master-e2e-metal-assisted-operator-capi/1517490862534168576

**Review by eranco74 — COMMENTED** [review]
I thought we wanted to place this in assisted-test-infra

**Comment by mkowalski** [comment]
InfraEnv from the manifest

```
 apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: 'assisted-infra-env'
  namespace: 'assisted-spoke-cluster'
spec:
  clusterRef:
    name: 'assisted-test-cluster'
    namespace: 'assisted-spoke-cluster'
  pullSecretRef:
    name: 'assisted-pull-secret'
  sshAuthorizedKey: 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC9cnjx0/JxoGM/0pKvSYR5Zgb80u2IFwDGZiSDg96vXLPhUW76s2F3lcw4QN1NHOxhtgzHqk+ADI1PpZh6BhtneqkZzNLliikUxDHA0XZyoSSbPHzS4nuuQV0Y26qD/ZsADPGX1uoybNZOVtuzDW6UEQCxUn1ofGPPglFoUwTT/Gs7Oz+0TNB509uVsekdPswMlGvV/1mwcKFkn0RUO/x67t6Wv4zA7fkQsO8mVfZT/IaoLUwPm+xVDc0Y1owrL4o79ah3r9NDrSLRpTv3VL0XEyBj1M+mxwtPRvmV/5O5WMarDgdesmH9iEuJg8I+9PAW4PeYN5ToqZyUOQlcLYegr/FhYpLdXBSZGeTHuBj/Vn7KhePI5q/fzKXDV2x47qU54fXQiHd5vWmpbEgzEJqdWdSBwCgX01O1iYa2kYdb5TB44FxZqnDLxG3cQGZfTfyRZic6jzjt8JeypGeUBaKxQgGhvzQQrMMjUM7f2Ari8YqWHpOiYb95ymUS8SWzX6M= root@ipi-ci-op-2x79sdpb-fe3d9-1517548543429054464' 
```

got created but the Image has never been created

```
infraenv.agent-install.openshift.io/assisted-infra-env created 
[...]
 + oc wait -n assisted-spoke-cluster --for=condition=ImageCreated --selector '' infraenv/assisted-infra-env --timeout=5m
error: timed out waiting for the condition on infraenvs/assisted-infra-env 
```

Debugging...

/hold

**Comment by mkowalski** [comment]
The issue was that InfraEnv must not contain clusterRef for CAPI and late-binding.

/test e2e-metal-assisted-operator-capi

**Comment by mkowalski** [comment]
```
 + oc get hostedcluster/assisted-test-cluster --selector= --namespace=assisted-spoke-cluster

NAME                    VERSION   KUBECONFIG                               PROGRESS   AVAILABLE   REASON                    MESSAGE
assisted-test-cluster             assisted-test-cluster-admin-kubeconfig   Partial    True        HostedClusterAsExpected   

[...]

 + oc scale nodepool/assisted-test-cluster -n assisted-spoke-cluster --replicas=1
nodepool.hypershift.openshift.io/assisted-test-cluster scaled 

[...]

 + oc extract -n assisted-spoke-cluster secret/assisted-test-cluster-admin-kubeconfig --to=-
# kubeconfig
+ export KUBECONFIG=/tmp/assisted-test-cluster-kubeconfig
+ KUBECONFIG=/tmp/assisted-test-cluster-kubeconfig
+ wait_for_object_amount node 1 10
+ object=node
+ amount=1
+ interval=10
+ namespace=
++ oc get node -n '' --no-headers
++ wc -l
No resources found 
```

Potentially there is a problem with extracting spoke cluster's kubeconfig. Debugging...

**Comment by mkowalski** [comment]
/test e2e-metal-assisted-operator-capi

Rolling with custom `http://quay.io/eranco74/hypershift:fix_kubeconfig` image

File: deploy/operator/capi/deploy_capi_cluster.sh [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+export SPOKE_API_VIP="${SPOKE_API_VIP:-}"
+export SPOKE_INGRESS_VIP="${SPOKE_INGRESS_VIP:-}"
+export ASSISTED_STOP_AFTER_AGENT_DISCOVERY="${ASSISTED_STOP_AFTER_AGENT_DISCOVERY:-false}"
+export ASSISTED_UPGRADE_OPERATOR="${ASSISTED_UPGRADE_OPERATOR:-false}"
+export USER_MANAGED_NETWORKING="${USER_MANAGED_NETWORKING:-false}"
+export SPAWN_NONE_PLATFORM_LOAD_BALANCER="${SPAWN_NONE_PLATFORM_LOAD_BALANCER:-false}"
+export ADD_NONE_PLATFORM_LIBVIRT_DNS="${ADD_NONE_PLATFORM_LIBVIRT_DNS:-false}"
+export LIBVIRT_NONE_PLATFORM_NETWORK="${LIBVIRT_NONE_PLATFORM_NETWORK:-ostestbm}"
+export LOAD_BALANCER_IP="${LOAD_BALANCER_IP:-192.168.111.1}"
+export HYPERSHIFT_IMAGE="quay.io/eranco74/hypershift:fix_kubeconfig" # FIXME(chocobomb)
```
[the remark this answers is not in the document]
Comment (mkowalski) [reply deploy/operator/capi/deploy_capi_cluster.sh:28]: Hypershift's broken at the moment and I'm just trying stuff around to see if this CI flow is going to pass once Hypershift upstream is fixed

File: deploy/operator/capi/deploy_capi_cluster.sh [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+function hypershift() {
+  podman run -it --net host --rm --entrypoint /usr/bin/hypershift -v $KUBECONFIG:/root/.kube/config -v $ASSISTED_PULLSECRET_JSON:/root/pull-secret.json -v /root/.ssh/id_rsa.pub:/root/.ssh/id_rsa.pub $HYPERSHIFT_IMAGE "$@"
+}
+
+echo "Installing HyperShift using upstream image"
+hypershift install --hypershift-image $HYPERSHIFT_IMAGE --namespace hypershift
+wait_for_pods "hypershift"
+
+echo "Creating HostedCluster"
+hypershift create cluster agent --name $ASSISTED_CLUSTER_NAME --base-domain redhat.example --pull-secret /root/pull-secret.json  --ssh-key /root/.ssh/id_rsa.pub --agent-namespace $SPOKE_NAMESPACE --namespace $SPOKE_NAMESPACE --control-plane-operator-image $HYPERSHIFT_IMAGE
```
[the remark this answers is not in the document]
Comment (mkowalski) [reply deploy/operator/capi/deploy_capi_cluster.sh:118]: As above, the final version won't have `--control-plane-operator-image`, but first we wait for Hypershift upstream to be fixed

**Review by javipolo — APPROVED** [review]

**Comment by mkowalski** [comment]
/unhold

We'll change the hypershift image from `quay.io/eranco74/hypershift:fix_kubeconfig` as soon as https://bugzilla.redhat.com/show_bug.cgi?id=2079177 is solved. For now let's have a test that works so that I can work on the `-gather` step

**Review by javipolo — APPROVED** [review]

**Pull request merged**